### DISCUSSION TITLE
Setting intel-oneapi-mkl as a package build dependency does not add the required paths to the environment variables

### DISCUSSION THREAD

**Discussion by dmentock** [spack/spack#30747] [Q&A] [answered]
I have a package that requires the `intel-oneapi-mkl` package as a build dependency. As opposed to regular packages, this package is not properly initialized in the respective environment variables when listing it as a package dependency:

```
depends_on('intel-oneapi-mkl%intel',type="build")
```
This only adds the root folders of those packages to the `CMAKE_PREFIX_PATH `variable and creates an entry for `intel-oneapi-mkl` in `MKLROOT`, but all the additions to `PATH`, `LD_LIBRARY_PATH `and `LIBRARY_PATH` are missing.
Loading the abovementioned spec with the `spack load` command works fine. I was not able to find a special rule in the documentation that excludes this package from being allowed as a build-dependency. 

I am also using the intel compiler for this package. Since I am compiling my own program using the `ifort `compiler, I need access to the `ifort `binary in `PATH` which also works when loading this compiler with the `spack load` command. When I compile my package by issuing `spack install <my_package>%intel` however, ifort is only loaded into the variables `SPACK_F77 `and `SPACK_FC` and becomes inaccessible to my installer that only looks for it in `$PATH`.

Issuing `depends_on('gcc@11.2.0',type="build")` in my package description adds the gcc compiler variables to `PATH` and `LD_LIBRARY_PATH`, which is why I assume that this is not a spack-comipler specific setting but rather directly related to the intel-oneapi-package.

Is there a clean solution that does not involve manually modifying the environment that allows me to access the intel-oneapi variables the same way I can when I am loading these packages via commandline?

@rscohn2

**Comment by rscohn2** [comment, marked as answer]
I believe spack load calls setup_run_environment: https://github.com/spack/spack/blob/51130abf8618c50361cdcd35be91f24ed422f607/lib/spack/spack/build_systems/oneapi.py#L106 which sources the oneapi vars.sh

Maybe I should make setup_dependent_build_environment do the same thing instead of only setting MKLROOT:
https://github.com/spack/spack/blob/51130abf8618c50361cdcd35be91f24ed422f607/var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py#L107
Could you see if it does what you want?

**Reply by dmentock** [reply]
I just added
```
env.extend(EnvironmentModifications.from_sourcing_file(
    join_path(self.component_path, 'env', 'vars.sh')))
```
to the `setup_run_environment` function of my local `intel-oneapi-compilers` and and `intel-oneapi-mkl` `package.py` files and was able to produce the desired behaviour. Would it be possible to add those changes to the public spack repository as well?

**Reply by rscohn2** [reply]
I changed the base class so ``setup_dependent_build_environment`` will source ``vars.sh`` for all oneapi packages in #30802.

**Comment by scheibelp** [comment]
> `depends_on('intel-oneapi-mkl%intel',type="build")`
> This only adds the root folders of those packages to the CMAKE_PREFIX_PATH variable and creates an entry for intel-oneapi-mkl in MKLROOT, but all the additions to PATH, LD_LIBRARY_PATH and LIBRARY_PATH are missing.

A package would have to be a link-type dependency to be added to `LD_LIBRARY_PATH`. The "build" dependency type means "this package is only needed during a build, and not after".

Does marking `intel-oneapi-mkl` as a "link" dependency resolve this?
